api/python/quilt3/data_transfer.py

```diff
@@ -171,6 +178,61 @@ def _build_unsigned_client(self):
         s3_client = self._build_client(lambda session: True)
         self._unsigned_client = s3_client
 
+    @staticmethod
+    def add_options_safely(params: dict, options: Optional[dict]):
+        """
+        Add options to the params dictionary safely.
+        This method ensures that the options do not overwrite existing keys in the params dictionary.
+        """
+        if options:
+            for key, value in options.items():
+                if key in params:
+                    raise ValueError(f"Cannot override key `{key}` using options: {options}.")
+                params[key] = value
+
+    def register_event_callback(self, event_name: str, callback: Optional[Callable]) -> None:
+        """Register a callback for S3 client events.
+
+        Args:
+            event_name: The name of the event to register for (e.g. 'creating-client-class')
+            callback: The callback function to be called when the event occurs, or None to remove the callback
+        """
+        if callback is None:
+            del self._event_callbacks[event_name]
+        else:
+            self._event_callbacks[event_name] = callback
+
+        # Force clients to re-register the callback on the new clients
+        self._standard_client = None
+        self._unsigned_client = None
+
+    def register_write_options(self, **kwargs):
+        """Register write options for S3 clients.
```

Comment:
I'd make users responsible for this

Reply:
So, I'm nervous about that.  It feels like we'd have to document all the internal APIs we call, or they'd have to use trial and error.  Or is that already true in general, and we should just include this as an example?

Reply:
Added as a doc example

Reply:
> So, I'm nervous about that. It feels like we'd have to document all the internal APIs we call, or they'd have to use trial and error. Or is that already true in general, and we should just include this as an example?

I just don't think we can make it work in general case e.g. because `SSECustomerKey` likely needs to be set for `complete_multipart_upload()` as well but that API doesn't support `ServerSideEncryption` and users may need to set different set of options for different buckets/keys
also it's quite hard to formulate why "write options" are set for `PutObject`, `CopyObject`, `CreateMultipartUpload` but not for `UploadPart` and `CompleteMultipartUpload`
in case they need to comply SCP I guess they need to figure out operations/options from the policy itself